src/librustc_codegen_llvm/back/lto.rs

```diff
@@ -111,29 +111,90 @@ fn prepare_lto(
-            for (name, data) in bytecodes {
+                .filter(|&(name, _)| {
+                    //eprintln!("- name {}", name);
+                    // njn: switch to using `looks_like_rust_object_file` here,
+                    // and make bitcode extraction brittle again
+                    name.ends_with(".o")
+                });
+            for (name, data) in object_files {
+                let name: &str = name;
+                let data: llvm::archive_ro::Child<'_> = data;
                 let _timer =
                     cgcx.prof.generic_activity_with_arg("LLVM_lto_load_upstream_bitcode", name);
-                info!("adding bytecode {}", name);
-                let bc_encoded = data.data();
-
-                let (bc, id) = match DecodedBytecode::new(bc_encoded) {
-                    Ok(b) => Ok((b.bytecode(), b.identifier().to_string())),
-                    Err(e) => Err(diag_handler.fatal(&e)),
-                }?;
-                let bc = SerializedModule::FromRlib(bc);
-                upstream_modules.push((bc, CString::new(id).unwrap()));
+                info!("adding bitcode from {}", name);
+                let object_data: &[u8] = data.data();
+
+                // njn: start by assuming fat object here, see what works and
+                // doesn't work
+
+                //eprintln!(
+                //    "  - object {} ({:?}, {})",
+                //    name,
+                //    object_data.as_ptr(),
+                //    object_data.len()
+                //);
+                match extract_bitcode(object_data) {
+                    Ok(bc) => {
+                        //eprintln!("    - bitcode {:?} {}", bc.as_ptr(), bc.len());
+                        if bc.len() > 0 {
+                            let bc = SerializedModule::FromRlib(bc.to_vec());
+                            upstream_modules.push((bc, CString::new(name).unwrap()));
+                        }
+                    }
+                    Err(e) => return Err(diag_handler.fatal(&e)),
+                }
             }
         }
     }
 
     Ok((symbol_white_list, upstream_modules))
 }
 
+fn extract_bitcode(obj: &[u8]) -> Result<&[u8], &'static str> {
+    //panic!("njn: hmm")
+
+    let mut data: *mut u8 = ptr::null_mut();
+    let mut len = 0;
+    let name = CStr::from_bytes_with_nul(b"njn: what here?\0").unwrap();
+    // njn: expand this unsafe zone?
+    let ok = unsafe {
+        // njn: use return value for the pointer, outparam for the length.
+        llvm::LLVMRustExtractBitcodeFromObject(
+            //cx,
+            obj.as_ptr(),
+            obj.len(),
+            name.as_ptr(),
+            &mut data,
+            &mut len,
+        )
+    };
+    if ok {
+        // njn: clean this up
+        //
+        // distinguish properly these cases:
+        /*
+        failure: nullptr, Err()
+        ok, but missing
+        ok, and zero-length
+        ok, and non-zero-length
```

Comment:
The bitcode starts with a magic number (https://www.llvm.org/docs/BitCodeFormat.html#magic-numbers) so it can't have a zero-length.